File: native/sensor-bridge/ios/DragonflySensorBridge/Sources/DragonflySensorBridge/SensorAdapter.swift
import Foundation

/// Every CGM vendor adapter conforms to this. The `Bridge` only ever talks
/// to adapters through this protocol so vendor SDK types stay isolated.
public protocol SensorAdapter: AnyObject {
    var vendor: GlucoseEvent.Vendor { get }
    var deviceName: String? { get }

    /// Becomes ready when the underlying SDK is initialized. The bridge
    /// will not start the adapter until ready emits `true`.
    var isReady: Bool { get }

    /// List sensors that the user could pair with. BLE scans for Dexcom;
    /// NFC tag prompt for Libre 1; the previously-paired Libre 2 device
    /// for live streaming. Implementations may return an empty list while
    /// scanning is pending.
    func discover() async throws -> [DiscoveredSensor]

    /// Bind to a specific sensor, requesting whatever auth/activation the
    /// vendor demands.
    func pair(_ sensor: DiscoveredSensor) async throws

    /// Begin emitting `GlucoseEvent`s through the supplied sink.
    func start(onEvent: @escaping (GlucoseEvent) -> Void) async throws

    /// Release the sensor and end the session.
    func stop() async
}

public struct DiscoveredSensor: Equatable, Hashable {
    public let id: String
    public let displayName: String
    public init(id: String, displayName: String) {
        self.id = id
        self.displayName = displayName
    }
}

public enum SensorAdapterError: Error {
    case notImplemented
    case notPaired
    case authorizationDenied
    case sdkUnavailable(String)
    case underlying(Error)
}

File: native/sensor-bridge/android/sensor-bridge/src/main/kotlin/com/dragonfly/sensorbridge/SensorAdapter.kt
package com.dragonfly.sensorbridge

/**
 * Boundary every CGM vendor adapter must implement. The Bridge only ever
 * talks to adapters through this interface so vendor SDK types stay
 * isolated.
 */
interface SensorAdapter {
    val vendor: GlucoseEvent.Vendor
    val deviceName: String?
    val isReady: Boolean

    /** List sensors the user could pair with. */
    suspend fun discover(): List<DiscoveredSensor>

    /** Bind to a specific sensor. */
    suspend fun pair(sensor: DiscoveredSensor)

    /** Begin emitting GlucoseEvents through the supplied sink. */
    suspend fun start(onEvent: (GlucoseEvent) -> Unit)

    /** Release the sensor and end the session. */
    suspend fun stop()
}

data class DiscoveredSensor(val id: String, val displayName: String)

sealed class SensorAdapterError(message: String) : RuntimeException(message) {
    object NotImplemented : SensorAdapterError("not implemented yet")
    object NotPaired : SensorAdapterError("adapter is not paired")
    object AuthorizationDenied : SensorAdapterError("authorization denied")
    class SdkUnavailable(reason: String) : SensorAdapterError(reason)
    class Underlying(cause: Throwable) : SensorAdapterError(cause.message ?: "underlying error")
}
